### DISCUSSION TITLE
Discord Webhooks and Auto Update

### DISCUSSION THREAD

**Discussion by TheDakk** [mbround18/valheim-docker#228] [Q&A]
Do the Discord webhooks only broadcast if there is actual update? Noticed it didn't broadcast anything during the auto update process.

**Comment by TjallingO** [comment]
Just to tack onto this: the only thing it seems to broadcast are manual stops / starts (for me, at least). More specifically, only these events have been broadcast thus far (and only when triggered manually):

```
Server Status: Stop Running
Server Status: Stop Successful
Server Status: Start Successful
```

**Reply by mbround18** [reply]
Anytime the container stops gracefully it should send stop and anytime you start the server it should send a start. 

Discord notifications are hard baked into the start and stop commands of Odin so there is no way to circumvent them unless you dont have the webhook environment set. My guess is, based on your feedback, I need to pass the WEBHOOK_URL to the cronjobs as well.

**Reply by mbround18** [reply]
See #234 which should have the fix for this, once this gets merged in give it a whirl <3

**Reply by TheDakk** [reply]
Thanks will check it out! Unrelated, but is there a way to change the discord notification colors? Like red if it’s stopped?

**Reply by mbround18** [reply]
This is something I am looking into adding :) As well as the actual notification header and message

**Reply by TheDakk** [reply]
[auto-update.log](https://github.com/mbround18/valheim-docker/files/6076502/auto-update.log)

So it auto updated this morning but it never sent the wehook notification for the startup:

Notification From: Your Server
Stop
Server Status: Stop Running
Notification From: Your Server
Stop
Server Status: Stop Successful

Also any chance to notify that it's running an auto update and alert if it found one or not?

**Reply by Devidian** [reply]
![image](https://user-images.githubusercontent.com/10413872/109819087-f9241580-7c33-11eb-9f15-e1bd17e5c917.png)

I also only got stop messages today - but the server was obviously still running and never stopped at all.

**Reply by ericbarrette** [reply]
I'm testing out the server since noon (EST) yesterday, and I also had the same behavior this morning.

There was a newer image pushed since then, that I pull this morning.  I do now receive the 3 notifications when the server update (Stop, Stops Success, Start), but now I do not get any notifications when I do `docker-compose up -d` and `docker-compose down`.  They don't even show up in the log.

**Reply by mbround18** [reply]
Huh, ill take a look at this

**Comment by dsme94** [comment]
I have noticed that despite the AUTO_UPDATE config in my docker-compose.yml file, the server does not automatically update, but does send messages into Discord at the cron time specified for a stop and start of the server. It doesn't ever actually update.

```
- AUTO_UPDATE=1
- AUTO_UPDATE_SCHEDULE="0 1 * * *"
```


I had to manually run this command:
```docker exec -it --user steam docker_valheim_1 bash ../scripts/auto_update.sh```

![image](https://user-images.githubusercontent.com/24738535/112633152-e797ed00-8e30-11eb-9bf7-7e4a4d994627.png)

From the screenshot above, the first lot of messages from Odin are the server stopping and starting at 1AM (when it should be updating, but no update messages...)
The second is from me running the docker exec command, and again no update messages...

**Reply by CosmicHorrorDev** [reply]
You're likely on an old version of the image. The server doesn't shutdown now unless there is an update available, so it shouldn't be shutting down daily anymore unless there is an update available every day

**Reply by Devidian** [reply]
It never did a shutdown but still displays the messages every day what is very confusing in my opinion. And fo me i set the auto update to 0500 local time but if the update comes ~ 1600 local time it is useless because i dont want to wait until next day for auto update.
![image](https://user-images.githubusercontent.com/10413872/112687044-2cdf0d80-8e77-11eb-9793-30f5e2b1134d.png)

Oh it only shows shutdown messages even if its not shutting down. Just checked last messages.

**Reply by CosmicHorrorDev** [reply]
Hmm, I just did some testing locally and wasn't able to reproduce this. I added in a line to the update script to indicate when it checks for updates and it shows that it checked for an update at the start of the hour, but no update was available so nothing else happened.

![image](https://user-images.githubusercontent.com/30302768/112698275-b7b40e00-8e5f-11eb-9845-e9e0ead37c3b.png)

What you seem to be saying is that it doesn't actually shut down, but still sends a notification for shutting down. Can you `exec` into your container

```bash
docker exec --user steam -it $CONTAINER_NAME bash
```

And check for an update manually along with posting the output of the update script

```bash
$ export DEBUG_MODE=1
$ odin update --check
$ echo $?
$ cat ~/scripts/auto_update.sh
```

And then post the results here? The auto update script should only shutdown and restart the server when an update is available, and the notification is only sent when `odin start` and `odin stop` is sent which shuts down and starts the server so a notification shouldn't go through unless the server is actually stopping.

Finally to address the timing situation. Since updating no longer requires shutting down the server you are totally free to check for updates more frequently like every hour. Just keep a couple things in mind:
- The valheim devs usually push the game update several hours before the server update, so if that becomes an issue for you, you can always set the client to not update automatically and then only update when the server update is actually available.
- Currently the server will shutdown if an update is found even if there are still players online. We're planning on making it so that the update will wait for all players to be off, and sending a discord notification that an update is available, but is waiting for an empty server, but I'm very busy with other stuff currently, so I don't have any spare time to work on that functionality.

**Reply by Devidian** [reply]
odin update is unknown it says, so i checked help:

```sh
steam@307bde172f21:~/valheim$ odin --help
odin 1.2.0
mbround18
Installs and Runs Valheim

USAGE:
    odin [FLAGS] [SUBCOMMAND]

FLAGS:
    -d, --debug      Sets the logger to log debug events.
    -r, --dry-run    Will output the commands that it intends to run.
    -h, --help       Prints help information
    -V, --version    Prints version information

SUBCOMMANDS:
    backup       Backups the current saves to a specifc location
    configure    Initializes Odin with its configuration variables.
    help         Prints this message or the help of the given subcommand(s)
    install      Installs Valheim with steamcmd
    notify       Sends a notification to the provided webhook.
    start        Starts Valheim
    stop         Stops Valheim
```

This is unexpected. So i checked my container - its using image id `c671a1a1b218` which is according to my `docker image list` 3 weeks old - but i always thought i had updated the image. Atleast i've pulled it because i also have latest (7 days old)
```sh
REPOSITORY                    TAG       IMAGE ID       CREATED        SIZE
mbround18/valheim             latest    0e26fdbcc52f   7 days ago     768MB
mbround18/valheim             <none>    c671a1a1b218   3 weeks ago    746MB 
```

So i just have updated again (pull image, docker-compose up -d) and its now using latest.

```sh
~/valheim$ odin update --check
[ODIN][DEBUG] - Debugging set to true
[ODIN][DEBUG] - Debug mode enabled!
[ODIN][DEBUG] - Launching update command...
[ODIN][INFO]  - Checking for updates
[ODIN][DEBUG] - Env var default 'ODIN_WORKING_DIR': '/home/steam/valheim'
[ODIN][DEBUG] - Env var default 'ODIN_WORKING_DIR': '/home/steam/valheim'
[ODIN][ERROR] - Failed to find steamcmd in path
[ODIN][ERROR] - Checking for script under steam user.
[ODIN][INFO]  - Executing: /home/steam/steamcmd/steamcmd.sh .....
[ODIN][INFO]  - Using steamcmd script at /home/steam/steamcmd/steamcmd.sh
[ODIN][DEBUG] - UpdateInfo {
[ODIN][DEBUG] -     current_build_id: "6394774",
[ODIN][DEBUG] -     latest_build_id: "6394774",
[ODIN][DEBUG] - }
[ODIN][INFO]  - No server updates found
~/valheim$ echo $?
0
```
So i think that was just the issue on my side, maybe someone else could find this information useful.
